### ISSUE TITLE
delete() crashes with ClassCastException for ID-Types other than String

### ISSUE THREAD

**Issue by aburmeis** [arangodb/spring-data#315]
With a `@Document` or `@Edge` annotated class having a `keyType=KeyType.uuid` and therefore an `@Id` annotated attribute of type `UUID` a call to `CrupRepository.delete()` (aka `SimpleArangoRepository.delete()`) will crash resulting in a `ClassCastException` as the code casts the identifier to `String` without any need:

```java
	public void delete(final T entity) {
		String id = (String) arangoOperations.getConverter().getMappingContext()
				.getRequiredPersistentEntity(domainClass).getIdentifierAccessor(entity).getRequiredIdentifier();
		arangoOperations.delete(id, defaultDeleteOptions(), domainClass);
	}
```

Just removing the cast and declaring id as `Object` should fix the issue:

```java
	public void delete(final T entity) {
		Object id = arangoOperations.getConverter().getMappingContext()
				.getRequiredPersistentEntity(domainClass).getIdentifierAccessor(entity).getRequiredIdentifier();
		arangoOperations.delete(id, defaultDeleteOptions(), domainClass);
	}
```

Even better also fix the Spring-Data contract of `CrupRepository.delete()`:
Throws: OptimisticLockingFailureException – [...] Also thrown if the entity is assumed to be present but does not exist in the database.

```java
	public void delete(final T entity) {
		Object id = arangoOperations.getConverter().getMappingContext()
				.getRequiredPersistentEntity(domainClass).getIdentifierAccessor(entity).getRequiredIdentifier();
                try {
		    arangoOperations.delete(id, defaultDeleteOptions(), domainClass);
		} catch (DocumentNotFoundException unknown) {
                    throw new OptimisticLockingFailureException("Non persistent entity", unknown)
		}
	}
```

**Comment by rashtao** [comment]
Thanks for reporting this @aburmeis.

AFAICS the related tests are mistakenly using generic types in entities and repositories, which are erased at runtime. See: https://github.com/arangodb/spring-data/blob/e84392af222d005ca9c08308198146c46a74b52e/src/test/java/com/arangodb/springframework/repository/AbstractTestEntityRepositoryTest.java.

The real tests for each type (e.g. for UUID https://github.com/arangodb/spring-data/blob/e84392af222d005ca9c08308198146c46a74b52e/src/test/java/com/arangodb/springframework/repository/IdTypeUUIDRepositoryTest.java) use at runtime persistent entities of type `IdTestEntity<Object>`.

Until this is tested and fixed, I would recommend using `String` for id and key fields.

**Comment by rashtao** [comment]
I fixed the tests in https://github.com/arangodb/spring-data/pull/316, but this lead to discover additional bugs. 

> Until this is tested and fixed, I would recommend using String for id and key fields.

**Comment by aburmeis** [comment]
Sorry @rashtao, I thought this would be a small fix not aware of the tests using not the right entity types.

**Comment by rashtao** [comment]
Fixed in #316, delete behavior throwing `OptimisticLockingFailureException` will be addressed in a separate PR.

 > Sorry @rashtao, I thought this would be a small fix not aware of the tests using not the right entity types.
 
No problem, thank you for spotting and reporting it!
